src/types/mappingproxy.rs

```diff
@@ -0,0 +1,614 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use super::PyMapping;
+use crate::err::PyResult;
+use crate::ffi_ptr_ext::FfiPtrExt;
+use crate::instance::Bound;
+use crate::types::any::PyAnyMethods;
+use crate::types::{PyAny, PyIterator, PySequence};
+use crate::{ffi, Python};
+
+/// Represents a Python `mappingproxy`.
+#[repr(transparent)]
+pub struct PyMappingProxy(PyAny);
+
+pyobject_native_type_core!(
+    PyMappingProxy,
+    pyobject_native_static_type_object!(ffi::PyDictProxy_Type),
+    #checkfunction=ffi::PyDictProxy_Check
+);
```

Comment:
I think rather than the ffi function (as discussed, it doesn't seem to exist in CPython), we should just have `fn PyDictProxy_Check` defined in this file.

```suggestion
    #checkfunction=PyDictProxy_Check
);

#[inline]
unsafe fn PyDictProxy_Check(op: *mut PyObject) -> c_int {
     ffi::PyObject_TypeCheck(op, std::ptr::addr_of_mut!(PyDictProxy_Type))
 }
```

Reply:
Also, it looks like subclassing `mappingproxy` is not allowed so we can probably just do `Py_IS_TYPE` here.

```python
>>> class F(types.MappingProxyType):
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: type 'mappingproxy' is not an acceptable base type
```

Reply:
Done, I changed the check function to the following:

```rust
#[inline]
unsafe fn dict_proxy_check(op: *mut ffi::PyObject) -> c_int {
    ffi::Py_IS_TYPE(op, std::ptr::addr_of_mut!(ffi::PyDictProxy_Type))
}
```